lib/data_pull.rb

```diff
@@ -0,0 +1,224 @@
+  # +#run(args: Array<String>, include_missing: Boolean) -> Result+
+  # @return [Class,nil]
+  def subtask(name)
+    {
+      'uuid-lookup' => UuidLookup,
+      'uuid-convert' => UuidConvert,
+      'email-lookup' => EmailLookup,
+      'profile-status' => ProfileStatus,
+    }[name]
+  end
+
+  def option_parser
+    @option_parser ||= OptionParser.new do |opts|
+      opts.banner = <<~EOS
+        #{$PROGRAM_NAME} [subcommand] [arguments] [options]
+
+        Example usage:
+
+          * #{$PROGRAM_NAME} uuid-lookup email1@example.com email2@example.com
+
+          * #{$PROGRAM_NAME} uuid-convert partner-uuid1 partner-uuid2
+
+          * #{$PROGRAM_NAME} email-lookup uuid1 uuid2...
+
+          * #{$PROGRAM_NAME} profile-status uuid1 uuid2...
+
+        Options:
+      EOS
+
+      opts.on('--help') do
+        config.show_help = true
+      end
+
+      opts.on('--csv') do
+        config.format = :csv
+      end
+
+      opts.on('--table', 'Output format as an ASCII table (default)') do\
+        config.format = :table
+      end
+
+      opts.on('--json') do
+        config.format = :json
+      end
+
+      opts.on('--[no-]include-missing', <<~STR) do |include_missing|
+        Whether or not to add rows in the output for missing inputs, defaults to off
+      STR
+        config.include_missing = include_missing
+      end
+    end
+  end
+
+  class UuidLookup
+    def run(args:, include_missing:)
+      emails = args
+
+      table = []
+      table << %w[email uuid]
+
+      uuids = []
+
+      emails.each do |email|
+        user = User.find_with_email(email)
+        if user
+          table << [email, user.uuid]
+          uuids << user.uuid
+        elsif include_missing
+          table << [email, '[NOT FOUND]']
+        end
+      end
+
+      Result.new(
+        log_message: "uuid-lookup, uuids: #{uuids.join(', ')}",
+        table:,
+      )
+    end
+  end
+
+  class UuidConvert
+    def run(args:, include_missing:)
+      partner_uuids = args
+
+      table = []
+      table << %w[partner_uuid source internal_uuid]
+      identities = AgencyIdentity.includes(:user, :agency).where(uuid: partner_uuids)
+
+      identities.each do |identity|
+        table << [identity.uuid, identity.agency.name, identity.user.uuid]
+      end
+
+      if include_missing
+        (partner_uuids - identities.map(&:uuid)).each do |missing_uuid|
+          table << [missing_uuid, '[NOT FOUND]', '[NOT FOUND]']
+        end
+      end
+
+      Result.new(
+        log_message: "uuid-convert, uuids: #{identities.map { |u| u.user.uuid }.join(', ')}",
+        table:,
+      )
+    end
+  end
+
+  class EmailLookup
+    def run(args:, include_missing:)
+      uuids = args
+
+      users = User.includes(:email_addresses).where(uuid: uuids)
+
+      table = []
+      table << %w[uuid email]
+
+      users.each do |user|
+        table << [user.uuid, *user.email_addresses.map(&:email)]
```

Comment:
Checking my understanding: The size of the array should remain consistent for the table, corresponding to the headers? But it looks like each row here could contain multiple email addresses.

Reply:
I forgot about the other output formats on this one. For CSV, it'll be fine, it's just a column with a blank header. The JSON will have problems, and I'm not sure how the ASCII table will handle

Reply:
I redid this so it has one row per email, and also adds a `confirmed_at` for additional debugging in 54279b2e48bac618b4915901f33dd20fefa71c8a
```
+--------------------------------------+-----------------+-------------------------+
| uuid                                 | email           | confirmed_at            |
+--------------------------------------+-----------------+-------------------------+
| 114a1840-b75c-4d5a-bb3d-6034ff5d5066 | aaa@example.com | 2023-04-06 23:08:35 UTC |
| 114a1840-b75c-4d5a-bb3d-6034ff5d5066 | bbb@example.com |                         |
| 114a1840-b75c-4d5a-bb3d-6034ff5d5066 | ccc@example.com |                         |
+--------------------------------------+-----------------+-------------------------+
```